libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -323,23 +544,246 @@ static MQTTStatus_t addRecord( MQTTPubAckInfo_t * records,
     return status;
 }
 
+/*-----------------------------------------------------------*/
+
 static void updateRecord( MQTTPubAckInfo_t * records,
                           size_t recordIndex,
                           MQTTPublishState_t newState,
                           bool shouldDelete )
 {
+    uint16_t packetId;
+
+    assert( records != NULL );
+
     if( shouldDelete == true )
     {
+        /* Mark the record as invalid. */
         records[ recordIndex ].packetId = MQTT_PACKET_ID_INVALID;
-        records[ recordIndex ].qos = MQTTQoS0;
-        records[ recordIndex ].publishState = MQTTStateNull;
     }
     else
     {
-        records[ recordIndex ].publishState = newState;
+        /* For QoS2 messages, in order to preserve the message ordering, when
+         * a PUBREC is received for an outgoing publish, the record should be
+         * moved to the last. This move will help preserve the order in which
+         * a PUBREL needs to be resent in case of a session reestablishment. */
+        if( newState == MQTTPubRelSend )
```

Comment:
What if instead of adding this logic here, add it to updateStateAck? That would keep this function simple and straightforward and move the reordering logic up to a higher level.
```
shouldDeleteRecord = ( ( newState == MQTTPublishDone ) || ( newState == MQTTPubRelSend ) ) ? true : false;
...
if( isTransitionValid == true )
{
    updateRecord( records, recordIndex, newState, shouldDeleteRecord );
    status = MQTTSuccess;
    if( newState == MQTTPubRelSend )
    {
        addRecord( records, MQTT_STATE_ARRAY_MAX_COUNT, packetId, MQTTQoS2, MQTTPubRelSend );
    }
}
```